### DISCUSSION TITLE
Rotation/orientation

### DISCUSSION THREAD

**Discussion by star69rem** [SnowflakePowered/librashader#5] [Q&A] [answered]
Is there documentation about the view matrix? I'm assuming that you have to pass it settings there to get images back that are rotated 90, 180, 270, etc.

I wanted to just rotate it myself, but I can't, because I end up having an input with a width that's less than the output that I want, and so I'm getting vertical scanlines because it probably thinks the image is coming in rotated.

**Comment by chyyran** [comment]
As of https://github.com/SnowflakePowered/librashader/commit/82ad516db0e597f03b5e9d0a0b45ad7390e0add8 Vulkan and DX11 run internal passes with an identity matrix so hopefully that should mimic RetroArch's behaviour.

The default MVP for use in all runtimes on the final pass is

```rust
pub static DEFAULT_MVP: &[f32; 16] = &[
    2f32, 0.0, 0.0, 0.0,
    0.0, 2.0, 0.0, 0.0,
    0.0, 0.0, 0.0, 0.0,
    -1.0, -1.0, 0.0, 1.0,
];
```
If you apply a rotation on that matrix it should work, Just pass a pointer to the `float[16]` in `*_frame`.

It remains to be seen if I want to provide a 'rotation' API option instead in radians or if just providing the final pass MVP is sufficient.

**Comment by star69rem** [comment]
A game's internal resolution can be 692x224, and if I correct the aspect ratio for a 90 degree rotation at 3840x1600, I want a 1600x666 resolution image. For some reason, though, when I do this, and then I rotate the output myself, I get vertical scanlines. 

It's really inconvenient to have to deal with another set of rotation code when I already have code to do it that works with my non shader path.

**Reply by chyyran** [reply]
The only mechanism that the shaders get for rotation is to tweak the MVP. You can give any `float[16]` matrix as an MVP, so you'll just have to account for the existing rotation. If it would help I could expose the MVP used for internal passes but otherwise I'm not sure what API shape you have in mind.

**Reply by chyyran** [reply]
Unfortunately I was mistaken when I was researching RetroArch's code base and `frame_direction` is actually meant for rewind.. So there really is nothing to pass to the shaders outside MVP.

**Reply by chyyran** [reply]
Even a `radians` based API would just be a high level API on top of matrix rotations.

**Reply by star69rem** [reply]
Sorry I probably wasn't articulating what I meant clearly. The problem I'm having is that I basically always want to pass in a normal orientation image, yet when I do that and rotate the output, I seem to get vertical scanlines for some reason. I think most of my problems are the result of some conflict between directxtk spritebatch and librashader.

**Reply by chyyran** [reply]
If you’re rotating the final output you probably need to provide an MVP such that the scanlines are vertical when applied to the normal orientation image. I don’t know how well the shaders will respond to that though. Why not rotate the input image and supply a properly rotated MVP to the shader?

**Reply by star69rem** [reply]
> If you’re rotating the final output you probably need to provide an MVP such that the scanlines are vertical when applied to the normal orientation image. I don’t know how well the shaders will respond to that though. Why not rotate the input image and supply a properly rotated MVP to the shader?

I have support for multiple non shader outputs, so I need to do my own rotation anyway for those. It's much simpler to be able to treat everything the same. I guess what I'm not getting is why a rotation that I apply after the shader frame is generated is giving me vertical scanlines. I would think that if i request a non rotated 1600x666 image, I would get a 1600x666 image with horizontal scanlines which I could then rotate myself, but it's not working that way. Like I said, I have the feeling that it's directxtk fighting with it sadly.

**Reply by chyyran** [reply]
The default MVP assumes that the input is unrotated, so it’ll apply horizontal scanlines to that. After you rotate it then the scanlines being vertical makes sense. You need to provide an MVP such that vertical scanlines are applied. 

I would think that just rotating the final MVP would be sufficient but if not I can expose the internal identity matrix as a parameter as well.

**Reply by star69rem** [reply]
> The default MVP assumes that the input is unrotated, so it’ll apply horizontal scanlines to that. After you rotate it then the scanlines being vertical makes sense. You need to provide an MVP such that vertical scanlines are applied.
> 
> I would think that just rotating the final MVP would be sufficient but if not I can expose the internal identity matrix as a parameter as well.


That's why I'm having problems. I'm passing in an unrotated image, expecting to get one back with horizontal scanlines, and then rotating myself, but what's happening is that I'm passing in an unrotated image and getting back one with vertical scanlines.

**Reply by chyyran** [reply]
Have you tried passing in a rotated MVP?

**Reply by star69rem** [reply]
> Have you tried passing in a rotated MVP?

No, because I want horizontal scanlines.

**Reply by chyyran** [reply]
If you pass in the rotated MVP you'll get vertical scanlines on the unrotated image, then when you rotate it they'll turn out horizontal.

**Comment by star69rem** [comment]
That doesn't make sense to me. I don't want vertical scanlines. I want horizontal scanlines and to rotate the whole image.

![image](https://user-images.githubusercontent.com/109876913/216825374-5f86af66-c11d-4f2d-865a-01bf9ecc77da.png)

**Comment by star69rem** [comment, marked as answer]
Please close this issue. What happened is that I had a super wide ultrawide image, and it just so happens when you have an image that thin rotated, just due to lack of resolution it looked like the scanlines were vertical when they weren't. When I tried rotating 4:3 aspect ratio modes, it looked okay. Sorry!
